### PR TITLE
Text parser and SWIG fixes

### PR DESCRIPTION
**Pull request by rpaciorek**
**LegacyTextParser fixes and enhancement**

- ~~restore support for \n in string as newline (lost while migration from DefaultRenderedStringParser to LegacyTextParser)~~ add `[br]` tag for line break
- fix `[color]` tag work while TextColour property is set to non white (especially when is black)
       - used `d_useModColour` variable to disable multiply setting colour and default colour (when set new color)
       - ~~user can use `[colour-modulate="true"]` `[colour-modulate="false"]` to manual manipulate `d_useModColour` (e.g. enabling color multiply, disabling it for image)~~
- implement `[aspect-lock]` tag
- try to load missing images from file
- add `[reset]` tag

**fix SWIG Python interface**
    
- replace RenderedStringParser by TextParser
- don't install OpenGL if not build it

### CODE REVIEW

File: cegui/src/text/LegacyTextParser.cpp [hunk 1]
```diff
@@ -106,16 +113,7 @@ bool LegacyTextParser::parse(const String& inText, std::u32string& outText,
...
-    d_bgColours = Colour(0x00000000);
-    d_font = nullptr;
-    d_padding = Rectf(0.f, 0.f, 0.f, 0.f);
-    d_imageSize = Sizef(0.f, 0.f);
-    d_vertFormatting = VerticalImageFormatting::BottomAligned;
-    d_outlineSize = 0.f;
-    d_underline = false;
-    d_strikeout = false;
-    d_styleChanged = true;
+    handleReset("full");
```
Comment (niello) [inline cegui/src/text/LegacyTextParser.cpp:116]: I don't like using interface string constants in the internal logic. Consider making a method `resetStyle(bool color, bool font ...)` or `resetColor`, `resetFont` etc, and calling them here and in `handleReset`. String constants should only appear in `handleReset` where they are parsed into internal commands.

File: cegui/src/text/LegacyTextParser.cpp [hunk 2]
```diff
@@ -163,9 +162,14 @@ bool LegacyTextParser::parse(const String& inText, std::u32string& outText,
 
                 if (escaped)
                 {
-                    // If it is not actually escaped, print the slash itself too.
+                    // We got \n, so make line break
+                    if (codePoint == 'n')
+                    {
+                        codePoint = '\n';
+                    }
```
Comment (niello) [inline cegui/src/text/LegacyTextParser.cpp:169]: This is a wrong place and at the same time the logic is inconsistent.
First, now you may render the text without a parser, and with your approach escaped characters will suddenly stop working. Also, every new parser will have to handle escaped characters. More robust would be to use another tag like [br] for this, as in HTML, making this explicitly belonging to a certain parser. Or use real newlines in the text, they are parsed by the common text rendering logic which is independent from the parser used.
Second, when you support \n, you should immediately think about \t, \r etc and why '\n' is an escape sequence and '\a' is literally "\a". Again, using [br] will solve this dilemma out of the box.

BTW, LegacyTextParser is called so because I wanted to implement a new text parser based on well known HTML and CSS text formatting tags and with a stack to enable intuitive nesting. But I really have no time for it now so you may do it if you are interested in text formatting, and I can help with reviewing and polishing the code. Another nice feature in my plan was a recursive argument substitution. If you are interested, we can discuss requirements and API.

File: cegui/src/text/LegacyTextParser.cpp [hunk 3]
```diff
@@ -216,6 +220,21 @@ bool LegacyTextParser::parse(const String& inText, std::u32string& outText,
     return true;
 }
 
+Sizef getRespectRatioSize(Sizef reqSize, const Sizef& orgSize) {
```
Comment (niello) [inline cegui/src/text/LegacyTextParser.cpp:223]: 1. Please follow the code style. A dash line before the function and a bracket on the new line when the function body begins.
2. Both Sizef args should be passed the same way. I prefer `const Sizef&` in this case. Using an argument as a temporary dest for internal calculations is not good, use local var instead.
3. Helper function in a module scope should be declared as `static`.

File: cegui/src/text/LegacyTextParser.cpp [hunk 4]
```diff
@@ -216,6 +220,21 @@ bool LegacyTextParser::parse(const String& inText, std::u32string& outText,
...
+	float orgRatio = orgSize.d_width / orgSize.d_height;
+	float reqRatio = reqSize.d_width / reqSize.d_height;
```
Comment (niello) [inline cegui/src/text/LegacyTextParser.cpp:225]: **Minor**: these two can be declared `const`. Not for astonishing performance boost but for clear statement that they are not going to change. This improves readability a bit.

File: cegui/src/text/LegacyTextParser.cpp [hunk 5]
```diff
@@ -216,6 +220,21 @@ bool LegacyTextParser::parse(const String& inText, std::u32string& outText,
...
+
+	if (reqSize.d_width == 0) {
+		reqSize.d_width  = reqSize.d_height * orgRatio;
+	} else if (reqSize.d_height == 0 || orgRatio > reqRatio) {
+		reqSize.d_height = reqSize.d_width  / orgRatio;
+	} else /*if (orgRatio < reqRatio)*/ {
+		reqSize.d_width  = reqSize.d_height * orgRatio;
```
Comment (niello) [inline cegui/src/text/LegacyTextParser.cpp:232]: 1. You have two of three branches equivalent. Please consider refactoring this code.
2. Your idea is not clear. A comment might help a reader to get it.
3. Do you absolutely need to strictly compare floats to zero? I see no division on checked values, so most probably you can get rid of this.

File: cegui/src/text/LegacyTextParser.cpp [hunk 6]
```diff
@@ -257,13 +276,25 @@ void LegacyTextParser::processControlString(const std::u32string& ctrlStr, std::
 
     if (valueValid && key == ImageTagName)
     {
+        const String val = ctrlStr.substr(valueStart, valueEnd - valueStart);
+        if (!ImageManager::getSingleton().isDefined( val )) {
```
Comment (niello) [inline cegui/src/text/LegacyTextParser.cpp:280]: Here and below - please follow the code style. We don't use spaces inside round brackets.

File: cegui/src/text/LegacyTextParser.cpp [hunk 7]
```diff
@@ -257,13 +276,25 @@ void LegacyTextParser::processControlString(const std::u32string& ctrlStr, std::
...
-            PropertyHelper<Image*>::fromString(ctrlStr.substr(valueStart, valueEnd - valueStart)));
+        auto element = std::make_unique<RenderedTextImage>( PropertyHelper<Image*>::fromString(val) );
         element->setColour(d_colours);
         element->setBackgroundColour(d_bgColours);
-        element->setSize(d_imageSize);
+        element->setUseModulateColour(d_useModColour);
+        if (d_imageAspectLock) {
+            element->setSize(getRespectRatioSize(
+                d_imageSize,
+                ImageManager::getSingleton().get( val ).getRenderedSize()
```
Comment (niello) [inline cegui/src/text/LegacyTextParser.cpp:293]: Redundant search in a manager. Consider accessing an image from an `element`, or better, making API like `element->setSize(d_imageSize, /*aspectLock=*/ true)`. This way you will not provide public access to internals.

**Review by niello — CHANGES_REQUESTED** [review]

File: cegui/include/CEGUI/text/LegacyTextParser.h [hunk 8]
```diff
@@ -60,7 +60,9 @@ namespace CEGUI
     - 'image-size' value is a CEGUI size property value.
     - 'image-width' value is a float.
     - 'image-height' value is a float.
-    - 'aspect-lock' value is a boolean (NB: this currently has no effect).
+    - 'aspect-lock' value is a boolean.
+    - 'reset' value is "full", "colour", "image" or "font" (reset color, font, image size, etc to default values).
+    - 'colour-modulate' value is a boolean (manual enable/disable color modulating with color from TextColour property of widget).
```
[the remark this answers is not in the document]
Comment (rpaciorek) [reply cegui/include/CEGUI/text/LegacyTextParser.h:65]: OK, if we don't want option to modulate color I remove 'colour-modulate' tag.

File: cegui/src/text/LegacyTextParser.cpp (hunk 2 above)
Comment (rpaciorek) [reply cegui/src/text/LegacyTextParser.cpp:169]: \n was backward compatibility (it was supported by DefaultRenderedStringParser), but I can change to [br]

File: cegui/src/text/LegacyTextParser.cpp (hunk 5 above)
Comment (rpaciorek) [reply cegui/src/text/LegacyTextParser.cpp:232]: Zero compare is used to allow `[image-width='100'][aspect-lock='true'][image='test.png']` without setting height for use auto height based on aspect ratio (and the same with height). Otherwise user code must be `[image-height='999999999'][image-width='100'][aspect-lock='true'][image='test.png']`.

And due to this we need three branches (order of test) - first check for zero (auto unlimited height/width), next compare real ratios.

I add some comments to this code.

File: cegui/src/text/LegacyTextParser.cpp (hunk 7 above)
Comment (rpaciorek) [reply cegui/src/text/LegacyTextParser.cpp:293]: OK, I make `RenderedTextImage::setSize(const Sizef& size, bool aspectLock)` and move `getRespectRatioSize` to `RenderedTextImage.cpp`

File: cegui/src/text/LegacyTextParser.cpp (hunk 2 above)
Comment (rpaciorek) [reply cegui/src/text/LegacyTextParser.cpp:169]: I think about html too (or `[]` tags wit multiple parameters), but no time too ... This PR is only moved stuff from my old custom parser (instead of ported it to new API). Main goal was image scaling based on aspect ratio and fix color for black text. For very advanced formatting I use CEF (and rendering to texture).

File: cegui/include/CEGUI/text/LegacyTextParser.h [hunk 9]
```diff
@@ -120,6 +126,8 @@ class CEGUIEXPORT LegacyTextParser : public TextParser
     void handleImageSize(const String& value);
     void handleImageWidth(const String& value);
     void handleImageHeight(const String& value);
+    void handleImageAspectLockTagName(const String& value);
```
[the remark this answers is not in the document]
Comment (rpaciorek) [reply cegui/include/CEGUI/text/LegacyTextParser.h]: fixed

File: cegui/src/text/LegacyTextParser.cpp (hunk 1 above)
Comment (rpaciorek) [reply cegui/src/text/LegacyTextParser.cpp]: fixed

File: cegui/src/text/LegacyTextParser.cpp (hunk 3 above)
Comment (rpaciorek) [reply cegui/src/text/LegacyTextParser.cpp]: fixed

File: cegui/src/text/LegacyTextParser.cpp (hunk 4 above)
Comment (rpaciorek) [reply cegui/src/text/LegacyTextParser.cpp]: fixed

File: cegui/src/text/LegacyTextParser.cpp (hunk 6 above)
Comment (rpaciorek) [reply cegui/src/text/LegacyTextParser.cpp]: fixed

**Review by niello — APPROVED** [review]
